File: tree/1008_construct_binary_search_tree_from_preorder_traversal/1008_construct_binary_search_tree_from_preorder_traversal.py
from typing import List


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

        
class Solution:
    def bstFromPreorder(self, preorder: List[int]) -> TreeNode:
        if len(preorder) == 0:
            return None
        leftList = []
        rightList = []
        root = TreeNode(preorder[0])
        for num in preorder[1:]:
            if num < preorder[0]:
                leftList.append(num)
            elif num > preorder[0]:
                rightList.append(num)
        root.left = self.bstFromPreorder(leftList)
        root.right = self.bstFromPreorder(rightList)
        return root

File: tree/1008_construct_binary_search_tree_from_preorder_traversal/1008_construct_binary_search_tree_from_preorder_traversal.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func bstFromPreorder(preorder []int) *TreeNode {
	if devide := len(preorder); devide > 0 {
		for i, val := range preorder {
			if val > preorder[0] {
				devide = i
				break
			}
		}
		return &TreeNode{
			Val:   preorder[0],
			Left:  bstFromPreorder(preorder[1:devide]),
			Right: bstFromPreorder(preorder[devide:]),
		}
	}
	return nil
}
